base/broadcast.jl

```diff
@@ -3,63 +3,207 @@
+Please see the Interfaces chapter of the manual for more information.
+"""
+rule(::Type{T}, ::Type{T}) where T        = T # homogeneous types preserved
+# Fall back to Unknown. This is necessary to implement argument-swapping
+rule(::Type{S}, ::Type{T}) where {S,T}   = Unknown
+# Unknown loses to everything
+rule(::Type{Unknown}, ::Type{Unknown})   = Unknown
+rule(::Type{T}, ::Type{Unknown}) where T = T
+# Precedence rules. Where applicable, the higher-precedence argument is placed first.
+# This reduces the likelihood of method ambiguities.
+rule(::Type{Nullable}, ::Type{Scalar})   = Nullable
+rule(::Type{Tuple}, ::Type{Scalar})      = Tuple
+rule(::Type{Bottom0d}, ::Type{Tuple})    = BottomVector
+rule(::Type{BottomArray{N}}, ::Type{Tuple}) where N  = BottomArray{N}
+rule(::Type{BottomArray{N}}, ::Type{Scalar}) where N = BottomArray{N}
+rule(::Type{BottomArray{N}}, ::Type{BottomArray{N}}) where N     = BottomArray{N}
+rule(::Type{BottomArray{M}}, ::Type{BottomArray{N}}) where {M,N} = _ruleba(longest(Val(M),Val(N)))
+# Any specific array type beats BottomArray. With these fallbacks the dimensionality is not used
+rule(::Type{A}, ::Type{BottomArray{N}}) where {A<:AbstractArray,N} = A
+rule(::Type{A}, ::Type{Tuple}) where A<:AbstractArray              = A
+rule(::Type{A}, ::Type{Scalar}) where A<:AbstractArray             = A
+
+## Allocating the output container
+"""
+    dest = similar(f, r::Broadcast.Result{ContainerType}, As...)
+
+Allocate an output object `dest`, of the type signaled by `ContainerType`,  for [`broadcast`](@ref).
+`f` is the broadcast operations, and `As...` are the arguments supplied to `broadcast`.
+See [`Broadcast.Result`](@ref) and [`Broadcast.rule`](@ref).
+"""
+Base.similar(f, r::Result{BottomArray{N}}, As...) where N  = similar(Array{eltype(r)}, indices(r))
+# In cases of conflict we fall back on Array
+Base.similar(f, r::Result{ArrayConflict}, As...)           = similar(Array{eltype(r)}, indices(r))
+
+## Computing the result's indices. Most types probably won't need to specialize this.
+indices() = ()
+indices(::Type{T}) where T = ()
+indices(A) = indices(combine_types(A), A)
+indices(::Type{Scalar}, A) = ()
+# indices(::Type{Any}, A) = ()
+indices(::Type{Nullable}, A) = ()
+indices(::Type{Tuple}, A) = (OneTo(length(A)),)
+indices(::Type{BottomArray{N}}, A::Ref) where N = ()
+indices(::Type{<:AbstractArray}, A) = Base.indices(A)
+
+### End of methods that users will typically have to specialize ###
 
 ## Broadcasting utilities ##
-# fallbacks for some special cases
-@inline broadcast(f, x::Number...) = f(x...)
+# special cases
+broadcast(f, x::Number...) = f(x...)
 @inline broadcast(f, t::NTuple{N,Any}, ts::Vararg{NTuple{N,Any}}) where {N} = map(f, t, ts...)
-broadcast!(::typeof(identity), x::Array{T,N}, y::Array{S,N}) where {T,S,N} =
-    size(x) == size(y) ? copy!(x, y) : broadcast_c!(identity, Array, Array, x, y)
+@inline broadcast!(::typeof(identity), x::AbstractArray{T,N}, y::AbstractArray{S,N}) where {T,S,N} =
```

Comment:
The signature relaxation tickled my memory: https://github.com/JuliaLang/julia/pull/19895. Not sure what the ramifications of this relaxation are now.